Move disassemble optimizations to optimization passes

yota9 opened the pull request:
The patch moves the shortenInstructions and nop remove to separate binary
passes. As a result when llvm-bolt optimizations stage will begin the
instructions of the binary functions will be absolutely the same as it
was in the binary. This is needed for the golang support by llvm-bolt.
Some of the tests must be changed, since bb alignment nops might create
unreachable BBs in original functions.

Vladislav Khmelevsky,
Advanced Software Technology Lab, Huawei

maksfb commented:
Thank you for this pull request. It's useful beyond just golang support.

yota9 commented on bolt/src/Passes/BinaryPasses.cpp:
```diff
@@ -1585,7 +1602,7 @@ void StripRepRet::runOnFunctions(BinaryContext &BC) {
 
 void InlineMemcpy::runOnFunctions(BinaryContext &BC) {
-  if (!BC.isX86())
+  if (!BC.isX86() || opts::NoInline)
```
Sure, I will create new PR now. It needed for golang, currently we cant afford inlining there :)

getianao commented:
That's a great patch and is useful for me. Can't wait to see the patch being merged.

yota9 commented:
Hello @maksfb ! It seems to be the test is incomplete. The makefile is wrong, e.g. to build the object file it needs to build bolt file first.  I've fixed that but I'm unable to build it, there are some compiler errors + there is no main function. I'm tried to bolt the executable that was in the archive and it is finished fine, but test seems to be want to read smth from stdio, so I'm not sure how to deal with it :)

maksfb commented:
Hi @yota9, my guess is that you cannot run the binary because of the dynamic dependencies. You should be able to see the difference in the binary output with and without your change though. Since you are just moving passes, I’d expect the binaries to be the same.

yota9 commented:
@maksfb I didn't investigate it deeply currently, but I see that the functions and basic blocks reorder might be different with/without patches. How do you think, is there a chance that due to instructions are remain in place with this patch in BF the profile reader/matcher  become more precise?

yota9 commented:
Hello @maksfb ! I think I've found where most of the differences are came from. First of all we don't need to skip these optimizations for non-simple functions (like we didn't do it before). But mostly important I think we will need to activate fixDoubleJumps optimization pass. There are some cases like 
```
BB1:
    < code >
    je BB4
BB2:
    nop
BB3:
    < code >
BB4:
    jmp BB3
```

So after NOP removal we will end up with empty BB. After fix-branches we will have double jmp BB1 -> BB2 -> BB3.
So with the second patch I'm temporary moved peephole passes up before fix-branches. Could you please check if the test works now adding `-peepholes=double-jumps` option? If yes I will refactor the code moving fix-branches to stand-alone pass and running it unconditionally, although I'm not sure will it fix the problem or not.

yota9 commented:
I've updated the patch, now the diff in text is about ~200 bytes, probably due to double-jumps option :)

maksfb commented:
Thanks, @yota9! Will give it a try.

yota9 commented:
@maksfb I've checked, the last bytes are changed by the fixBranches call, it inverts the cond jumps in a few places, I don't think it's a big deal. Also we should probably preserve branch information in replaceSuccessor call in fixDoubleJumps. Anyway I hope it will work now

maksfb commented:
@yota9, still got a crash.

How do you use the test case I've uploaded? You have to add `llvm-bolt` to your `PATH` and then run `make`.

> It seems to be the test is incomplete. The makefile is wrong, e.g. to build the object file it needs to build bolt file first.

The object file is going to be built using the BOLTed compiler `cc1plus.bolt`. `%.bolt` rule in the `Makefile` will produce that compiler.  `cc1plus.bolt` is the one that is crashing for me with the change.

> I've fixed that but I'm unable to build it, there are some compiler errors + there is no main function. I'm tried to bolt the executable that was in the archive and it is finished fine, but test seems to be want to read smth from stdio, so I'm not sure how to deal with it :)

Oh. I guess you were trying to run BOLT on `bf_gcc`? `bf_gcc.cpp` should be passed to `cc1plus.bolt` using the options in the `Makefile`.

yota9 commented:
@maksfb Oh I see, thanks for your explanation! I'm having Segmentation fault now too

yota9 commented:
@maksfb The problem is somewhere in frame optimization pass, it moves some of the instructions , e.g. push in the prologue to the epilogue for some reason. I will try to continue investigations

yota9 commented:
Hello @maksfb ! I've found the problem, it is located in ShrinkWrapping pass. I'm not 100% sure why my patches are causes this problem to appear, but the idea is that in some cases (e.g. `_ZL12process_usesP8df_ref_di`) the r12 register is saved in restored in function prologue/epilogue, but it is not used in the function. The pass finds that the best insertion point for the push instruction is located in the same place with the pop. But it emits it in the wrong order, first pop, the push. Currently I just changed it to always emit push instructions first. So now it works fine. I'm not sure if it is the best solution, we might eliminate such instructions, but I since this case is extremely rare I think we can use it as-is, without extra logic.
If this solution is appropriate I will refactor the patches and squash them together. Please also tell me what do you think about enabling fixDoubleJumps pass by default as I wrote previously. Thank you!

maksfb commented:
Hi @yota9, thank you for the investigation and for the fix. @rafaelauler, could you comment on the shrink-wrapping pass?

Do you know what is different about the function where the bug surfaced? Is it a different CFG or a different profile with your change?

I believe we have to change the profile-reading logic starting here: https://github.com/facebookincubator/BOLT/blob/61a2062542cd7f015cde0e0ff06abe305993943d/bolt/lib/Profile/DataReader.cpp#L744-L748

yota9 commented:
@maksfb I've updated the last patch, it is the stable_sort the place where the error occurs I think. We need to place stores first, and the conditions for A and B were messed I think

maksfb commented:
With the introduction of `nop` basic blocks, it makes sense for me to run `fixDoubleJumps` right after `removeNops`. I don't think you need to invoke `fixBranches()`. What was the reason you've added the call to it?

I don't have the details, but we've run into a problem with `fixDoubleJumps()` on AArch64 in the past. It was something related to tail calls.

rafaelauler commented:
Thanks @yota9 for fixing the bug in shrink wrapping! Changes on the shrink-wrapping sorting look good to me.

I've created a testcase for your change, could you please put it together with your change? Here's the testcase: https://pastebin.com/nfTZKCpQ

Regarding your question, you're right, I think we would need to reorder the stores as well, but I haven't been able to create a test case that triggers a bug with that yet. I'll take a look at that.

Also, once this is ready, could you squash all your changes in a single commit, or put each commit in a separate PR? Thanks!

yota9 commented:
@rafaelauler Thank you for the test case! It's in #251 
@maksfb I didn't check it, I just done it due to the jump instruction should change it's target, I have another idea of fixing this anyway. But back to this patch, it seems to be that the problem have solved, I'm not sure if I need to change DataReader, it seems to be working fine. As I saw the changes in that function are caused by not removed double jmp for some reason, however I'm not 100% sure to be honest, but the basic blocks stat seems to be right. So I hope that after refactoring it will be ready to go.

yota9 commented:
Hello @maksfb @rafaelauler ! I've updated the nop removal pass, checking that BB is empty we may change the successor in the predecessors, than the EliminateUnreachableBlocks pass will remove such BBs. 
But one test is failing - bb_with_two_tail_calls. I'm not sure what this test is checking, but it seems to be although jmp tailcall is not merged (I'm not sure why it have to) in the previous BB, the function is optimized even better:
```
Before:
Binary Function "_start(*2)" after simplify-conditional-tail-calls
.LBB00 (1 instructions, align : 1)
  Entry Point
  Exec Count : 0
  CFI State : 0
    00000000:   ja      .Ltmp0
  Successors: .Ltmp0 (mispreds: 0, count: 0), .LFT0 (mispreds: 0, count: 0)
  CFI State: 0

.LFT0 (2 instructions, align : 1)
  Exec Count : 4
  CFI State : 0
  Predecessors: .LBB00
    00000002:   jae     e # TAILCALL  # CTCTakenCount: 0
    00000004:   jmp     e # TAILCALL 
  CFI State: 0

.Ltmp0 (0 instructions, align : 1)
  Exec Count : 0
  CFI State : 0
  Predecessors: .LBB00
  Successors: .Ltmp1 (mispreds: 0, count: 0)
  CFI State: 0

.Ltmp1 (1 instructions, align : 1)
  Exec Count : 4
  CFI State : 0
  Predecessors: .Ltmp0
    00000009:   jmp     e # TAILCALL 
  CFI State: 0

```

```
Now:
Binary Function "_start(*2)" after simplify-conditional-tail-calls
.LBB00 (1 instructions, align : 1)
  Entry Point
  Exec Count : 0
  CFI State : 0
    00000000:   ja      e # TAILCALL  # CTCTakenCount: 0
  Successors: .LFT0 (mispreds: 0, count: 0)
  CFI State: 0

.LFT0 (1 instructions, align : 1)
  Exec Count : 4
  CFI State : 0
  Predecessors: .LBB00
    00000002:   jae     e # TAILCALL  # CTCTakenCount: 0
  Successors: .Ltmp1 (mispreds: 2, count: 4)
  CFI State: 0

.Ltmp1 (1 instructions, align : 1)
  Exec Count : 4
  CFI State : 0
  Predecessors: .LFT0
    00000004:   jmp     e # TAILCALL 
  CFI State: 0

```

The result seems to be expected, but replacing the test conditions to the new one may change the test meaning, and since I'm not quite understand what this test is checking I would like to ask for your help. Thank you!

maksfb commented:
Hi @yota, any update on `fallthrough-to-noop.test` failure?

yota9 commented:
Hello @maksfb. Sorry I didn't have time to check it yesterday, I will back to this soon :)

maksfb commented:
@yota9, we need to provide "backwards compatibility" for .fdata produced before this change. `fallthrough-to-noop.test` fails because the destination for the fall-through is in the middle of the basic block. We can check that it's right after the nop, but then we need the `Size` annotation.

yota9 commented:
@maksfb The size for the noop Instruction you mean? The golang pass already adds it buy the way :)

maksfb commented:
Yes, that's what I meant since you've mentioned it earlier.

yota9 commented:
Hello @maksfb ! Thank you for your comments, I've fixed the test, although as you said the problem was only with "backwards compatibility", that is why I didn't face this issue before, since we don't have such needs :) Anyway seems to be everything works now as expected

maksfb commented:
Hi @yota9, thank you for the quick fix. We have more .fdata compatibility issues and will post a test case soon.

yota9 commented:
@maksfb Is it the test problem or you need that for your production? To be honest I'm not really like the code I've added today, since new profile will work just fine. It is a question of how many issues and extra code we will need to add to support backward compatibility.  I mean if we are speaking about production env, than probably you are getting the profile each time the software is changed, since the old profile is no longer valid obviously.. Do you have some services that are rarely changed and you use the same profile each time? Maybe we should add profile versioning or something to raise an error in case the version mismatch, what do you think?

rafaelauler commented:
@yota9  Can you try these tests with your patch? https://github.com/rafaelauler/bolt-tests

I've put some instructions in the README regarding how to run these tests. Let me know if it works for you.

yota9 commented:
@rafaelauler I've managed to reproduce it, thank you! It will need some time to figure out for all 8 tests :(( As I understand currently you have some problems only with aarc64 and the x86 is OK?

rafaelauler commented:
@yota9 we have some large x86 tests failing too, I'll try to add them to the repo when I have some time.

yota9 commented:
Hello @maksfb @rafaelauler ! After I've replaced  ReorderBasicBlocks to be after EliminateUnreachableBlocks 2 tests from the repo above are almost works - only the "executed instructions" counter is a little bit differs and the "UCE removed 0 blocks and 0 bytes of code" should be changed/removed since we are removing empty BBs now. As I see most of the other tests are using non-LBR profiles. And as I understand readSampleData doesn't use offset annotation. So when previosly we removed the nops the offsets were shifted and the samples might started point out on the wrong BBs. Now when the nops are in place the statistics seems to be calculated right, and the rest of the changes in binaries are came from BB reordering and hot-cold splitting (e.g. I see that some of the blocks previously was put in the cold section, not it is in hot and vice-versa). What do you think about it? Thank you!

yota9 commented:
Hello @maksfb @rafaelauler ! Any thoughts on the investigation above? Thank you!

maksfb commented:
@yota9, apologies for the delay. Will take a look later today.

maksfb commented:
> As I see most of the tests are using non-LBR profiles.

Which tests? I believe most of our tests should be using LBRs.

maksfb commented:
> > As I see most of the tests are using non-LBR profiles.
> 
> Which tests? I believe most of our tests should be using LBRs.

Oh, AArch64?

yota9 commented:
@maksfb Yes, take a look at the tests @rafaelauler gave about for aarch64. The profiles are mostly (all?) non-LBR

maksfb commented:
> @maksfb Yes, take a look at the tests @rafaelauler gave about for aarch64. The profiles are mostly (all?) non-LBR

Right, for AArch64 that's all we have. Please add LBR tests for AArch64 if you have them.

We have x86-64 tests failing as well, they are just not in the repo yet. They use LBR, but likely the offsets are off as well.

Sadly, we need to keep backwards profile compatibility at least for some time, and then we will have to update all our tests if we are going to remove it.

Did you think about removing empty blocks and fixing double jumps right after RemoveNops pass? Or is there a reason not to do it for golang?

yota9 commented:
@maksfb I've updated the pass, please take a look, it is now making empty BBs unreachable and they will be removed by EliminateUnreachableBlocks pass later, do you think we need to elliminate them right in the pass? 

>> Right, for AArch64 that's all we have. Please add LBR tests for AArch64 if you have them.

>> Sadly, we need to keep backwards profile compatibility at least for some time

If I understand it correctly the problem is that for non-LBR tests we won't be able to make absolute compatibility. As I said in non-LBR mode the offset annotations are ignored, so previously when the nops were removed the offsets in fdata samles might started to point on the wrong BB, which is located furher, since the offset was shifted. So now with NOPs it is pointing to the right BBs, but the  BB reordering and hot-cold splitting might result in the other binary layout, since some of the previously HOT bbs might be cold now due to the nops.

maksfb commented:
> @maksfb I've updated the pass, please take a look, it is now making empty BBs unreachable and they will be removed by EliminateUnreachableBlocks pass later, do you think we need to elliminate them right in the pass?

The sooner in the pass manager we eliminate empty BBs the better. It will put less burden on later optimization passes. We only need nops and containing BBs for accurate byte-level representation of the input and for special cases where nops should be preserved. Right?

> If I understand it correctly the problem is that for non-LBR tests we won't be able to make absolute compatibility. 

Yes, sorry I wasn't clear, the backwards compatibility is needed for LBR profiles only. For AArch64 we need to update tests; and I cannot do it since I have limited access to such hardware. If you can update them or update with LBR profiles that's all we need. If you don't do this in this PR we'll have broken AArch64 tests.

yota9 commented:
>>  We only need nops and containing BBs for accurate byte-level representation of the input and for special cases where nops should be preserved. Right?
Sure. Although I don't think the empty unreachable BBs will make much of the difference but it is not a big deal to remove them right there.

>> If you don't do this in this PR we'll have broken AArch64 tests.

This is FBs internal tests, they are not in the repo, @rafaelauler loaded them is separate one :) 

Currently all the X86 LBR tests I have are passed

maksfb commented:
> This is FBs internal tests, they are not in the repo, @rafaelauler loaded them in separate one :)

The reason we didn't put them in the main repo is that we don't want medium/large binaries in the main repo. The external repo will contain binary tests.

> Currently all the X86 LBR tests I have are passed

I see a couple of test failures in CI (other than the meta test).

yota9 commented:
@maksfb Sorry, my fault. Updated + erase invalid BBs in remove-nops pass.

yota9 commented:
@maksfb All the tests (except meta) are passed

maksfb commented:
Great. Thank you, @yota9. I will likely have to add more profile backwards compatibility fixes for our internal tests before we add these tests to the external binary repo.

Meanwhile, think about adding AArch64 LBR tests. The repo added by @rafaelauler is at a temporary location. Once we integrate into the monorepo, we would prefer moving this binary repo under llvm umbrella (similar to llvm-test-suite repo).

yota9 commented:
@maksfb Great, I hope we will be able to submit this in the nearest time :)
>> Meanwhile, think about adding AArch64 LBR tests. The repo added by @rafaelauler is at a temporary location.

OK, we will add the task to replace the fdata in these tests to LBR ones to our backlog, we will handle this later.

yota9 commented:
Hello @maksfb @rafaelauler ! I'm aware about aarch64 tests, our team will handle it, is there something else I can help you to merge this PR? Thank you!

maksfb commented:
Hi @yota9, I'm currently working on fixing crashes related to this PR, adding backwards compatibility, and making sure all nops are removed.

yota9 commented:
Thank  you @maksfb ! I hope it will go smoothly :)

yota9 commented:
Hello @maksfb @rafaelauler ! @ElvinaYakubova tried to intrument the only one left arm test from the repo above to get the LBR profile but the test is compiled without emit-relocs. Also the inputs for the test is needed to get the new profile. Thank you!

yota9 commented:
Hello @maksfb @rafaelauler ! Any updates on this? Thank you

maksfb commented:
@yota9, I think I resolved all issues but one that I'm currently tackling.

maksfb commented:
This PR should be integrated after the next sync. I made some changes to the original PR code, let me know if they look good you. Thanks again.

yota9 commented:
Hello @maksfb ! Great news, will wait for this, thank you!

maksfb commented:
The PR went in, but it was marked as authored by me. We'll fix it.

yota9 commented:
Hello @maksfb . Overall LGTM, thank you!
Also I would like to ask you about current PRs and golang support. As I understand you are busy with the llvm integration and there is no reason to create MR with golang support now? 
About current PRs:
The only one must-have golang support patch is #239 , if it is possible I would like this PR to submit separately from golang support. The #267 #268 are also needed to support update-debug-sections, but they are not golang - specific patches. The #230  I will probably close for now.
Please also take a look to the couple of lines patch #258 - it memory corruptions in BOLT :)

maksfb commented:
@yota9, I will prioritize reviewing the patches. But, yes, we are busy with the merge process. Do you want to add yourself as the code owner of aarch64 backend to bolt/CODE_OWNERS.TXT? We also need more aarch64 binary tests in the binary repo (https://github.com/rafaelauler/bolt-tests).

yota9 commented:
Hello @maksfb! Great, thanks!
>> Do you want to add yourself as the code owner of aarch64 backend to bolt/CODE_OWNERS.TXT?

I will be glad to do this, thank you!

>> We also need more aarch64 binary tests 

Most of our tests are golang compiler tests + internal services to be honest. Anyway we will think about what tests we can add there with the LBR profiles for ARM.
